cpp/src/arrow/type.cc

```diff
@@ -686,6 +686,48 @@ std::shared_ptr<Schema> schema(std::vector<std::shared_ptr<Field>>&& fields,
   return std::make_shared<Schema>(std::move(fields), metadata);
 }
 
+Status UnifySchemas(const std::vector<std::shared_ptr<Schema>>& schemas,
+                    std::shared_ptr<Schema>* out) {
+  if (schemas.empty()) {
+    return Status::Invalid("Must provide at least one schema to unify.");
+  }
+
+  std::unordered_map<std::string, std::shared_ptr<Field>> field_name_to_field;
```

Comment:
Do you want to make a best effort to be order preserving? Rather than use just a map, you could use a string-to-index map and then a vector containing the ordered fields as observed.

```
std::u_m<string, int> field_name_to_index;
std::vector<sp<Field>> all_fields;
```

So if a new field is observed it will get tacked onto the end rather than have them all jumbled

Reply:
That was a question I raised at the beginning.

To what extent should the order be preserved or be deterministic?

The current implementation does not preserve the order, and is not deterministic either (if you call it multiple times, depending on the unordered_map implementation, you might get different ordering).

Do we want both, or either, or None?

To the use cases I'm interested in, ordering / determinism does not matter.

Reply:
I made the change to preserve the ordering.